prow/cmd/sinker/main.go

```diff
@@ -532,3 +573,80 @@ func podNeedsKubernetesFinalizerCleanup(log *logrus.Entry, pj *prowapi.ProwJob,
 
 	return false
 }
+
+func getDuplicates(l *logrus.Entry, pjs *prowapi.ProwJobList) (result []string, err error) {
+	start := time.Now()
+	defer func() {
+		l.WithField("num_duplicates_found", len(result)).WithField("getDuplicates duration", time.Since(start)).Debug("getDuplicates finished")
+	}()
+	type duplicationKey struct {
+		job string
+		// pointer types are only equal
+		// when they point to the same
+		// address, so we have to serialize
+		// the value of these fields
+		refs      string
+		extraRefs string
+		state     prowapi.ProwJobState
+	}
+	duplicates := map[duplicationKey][]string{}
+
+	var key duplicationKey
+	for _, pj := range pjs.Items {
+		if pj.Spec.Type == prowapi.PeriodicJob || !pj.Complete() {
+			continue
+		}
+
+		key.job = pj.Spec.Job
+		key.refs, err = comparableRefsString(pj.Spec.Refs)
+		if err != nil {
+			return nil, err
+		}
+		key.extraRefs, err = comparableExtraRefsString(pj.Spec.ExtraRefs)
+		if err != nil {
+			return nil, err
+		}
+		key.state = pj.Status.State
+
+		duplicates[key] = append(duplicates[key], pj.Name)
+	}
+
+	for _, duplicateSlice := range duplicates {
+		if len(duplicateSlice) == 1 {
+			continue
+		}
+		result = append(result, duplicateSlice[1:]...)
+	}
+
+	return result, nil
+}
+
+func comparableExtraRefsString(extra []prowapi.Refs) (string, error) {
+	if len(extra) == 0 {
+		return "", nil
+	}
+	if len(extra) == 1 {
+		return comparableRefsString(&extra[0])
+	}
+	b := &strings.Builder{}
+	for _, ref := range extra {
+		res, err := comparableRefsString(&ref)
+		if err != nil {
+			return "", err
+		}
+		b.WriteString(res)
+	}
+
+	return b.String(), nil
+}
+
+func comparableRefsString(r *prowapi.Refs) (string, error) {
```

Comment:
Surely there's something from `cmp` that gives us this comparison without a marshal? Or, can't we get a `DeepEqual` generated for us by client-go stuff since this is part of the CR? Or, barring that, just write an equality func by hand and a fuzzing test for it so it stays up to date? This seems expensive to be putting into such a hot callpath.

Reply:
This is a map, maps fundamentally need a key without providing any input, otherwise they would always need to iterate over all elements on fetch and iterate over all possible keys when you insert a new element.

I thought about using a combination of a map and a slice by using  a map of jobnames with a two dimensional job array, first dimension is the unique refs, then iterating and using a comparison func to compare. But this is definitely a lot more complex and might not even be faster.

Reply:
Sorry, I'm not following. Why are we choosing to do JSON serialization and string comparison over some dynamic comparison from e.g. `cmp` or a comparison func written for this purpose?

Reply:
We group all prowjobs that have the same (.spec.name, spec.refs, spec.extraRefs, status.State). Using a comparison func would require us to call the comparison func for all groups for all prowjobs, unless we find a match earlier. Since we will have a lot of groups this can be assumed to be more expensive than what we currently do. Additionally, it is more complicated. So for now, this is the simplest solution. If we end up finding out that we need a more efficient approach, we will have to profile alternate approaches.